C++: Handle constant variable accesses in SimpleRangeAnalysis.qll

MathiasVP opened the pull request:
The PR fixes #3009. The problem was that a `VariableAccess` to a static const member variable does not have a `getValue()` result, and this meant that `analyzableExpr` in `SimpleRangeAnalysis.qll` wasn't matched for that variable access.

I couldn't verify this problem with a simple qltest, so I assume that this only happens when the initialization happens in a different compilation unit.

Another fix for this problem might be to override `getValue()` in `Access.qll`, but that feels like quite an intrusive change.

jbj commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:
```diff
@@ -91,13 +91,30 @@ private float wideningUpperBounds(ArithmeticType t) {
   exists(exprMinVal(e)) and
   (
-    exists(e.getValue().toFloat()) or
+    exists(getValue(e).toFloat()) or
```
This is just one of several calls to `getValue`. Don't the others need to change too?

MathiasVP replied:
Good point. I totally forgot about the rest!

jbj commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:
```diff
@@ -91,13 +91,30 @@ private float wideningUpperBounds(ArithmeticType t) {
+    exists(VariableAccess access, Variable v |
+      e = access and
+      v = access.getTarget() and
+      v.getType().isConst() and
```
```suggestion
      v.getUnderlyingType().isConst() and
```

This should make sure we handle typedefs.

MathiasVP replied:
Oh, right! Good catch.

jbj commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll:
```diff
@@ -91,13 +91,30 @@ private float wideningUpperBounds(ArithmeticType t) {
+      e = access and
+      v = access.getTarget() and
+      v.getType().isConst() and
+      result = getValue(v.getAnAssignedValue())
```
I don't think this value will always be unique. What if you have `const int C = 1;` in one file and `const int C = 2;` in another? And then `extern const int C;` in a third, which accesses `C`. Here I'm thinking C, not C++. IIUC, constants have external linkage in C and can be used like this, but I haven't checked.

You can use the trick of requiring that `result = min(...) and result = max(...)` to ensure there's one unique result. Or you can find lower and upper bound separately with `min` and `max` respectively.

MathiasVP replied:
> IIUC, constants have external linkage in C and can be used like this, but I haven't checked.

I was not aware of this difference between C and C++! That should indeed be handled.

MathiasVP replied:
I chose to keep the `getValue()` predicate as is (to avoid the need for `monotonicAggregates`), and instead do the `result = min(...) and result = max(...)` trick in a separate "top level" predicate. The tuple count looked fine on the projects I tested.

jbj replied:
I think I've given you bad advice here. It looks like the callers of `getValue` are able to tolerate multiple results. The only caller for which multiple values might matter is `getTruncatedLowerBounds` (and its twin for upper bounds). Most callers of `getTruncatedLowerBounds` are in the main recursion, where multiple results are tolerated as long as they are not numerous enough to kill performance. The value ends up in `lowerBound`, where a `min` is applied.

MathiasVP replied:
I see. I've reverted this specific change now. Do you think we need to add a `rank[1..10]` expression (or something similar) to ensure that there aren't too many results?

jbj replied:
No, I think it's good now. If the results should explode on some particular codebase, then I'd like to understand why that happens before we choose a way to mitigate it.

MathiasVP commented:
> If you can't write a simple qltest for it then how about a qltest with two C files?

I've done this now. Didn't know this was possible! The new testcase I added resulted in an alert:
```
Diff:
--- expected
+++ actual
@@ -2,3 +2,4 @@
 | test.c:9:14:9:18 | ... > ... | Comparison between $@ of type char and $@ of wider type int. | test.c:8:7:8:7 | c | c | test.c:7:17:7:17 | x | x |
 | test.c:14:14:14:18 | ... < ... | Comparison between $@ of type short and $@ of wider type int. | test.c:13:8:13:8 | s | s | test.c:12:17:12:17 | x | x |
 | test.c:65:14:65:18 | ... < ... | Comparison between $@ of type short and $@ of wider type int. | test.c:64:8:64:8 | s | s | test.c:63:17:63:17 | x | x |
+| test.c:77:13:77:24 | ... < ... | Comparison between $@ of type short and $@ of wider type int. | test.c:76:8:76:8 | s | s | test2.c:1:11:1:18 | const256 | const256 |
```
which is now removed with this PR.
